sopel/plugins/callables.py

```diff
@@ -0,0 +1,586 @@
+"""Plugin objects' definition.
+
+.. versionadded:: 8.1
+
+The core of a plugin consists of several objects, either plugin callables or
+plugin jobs. These objects are represented by instances of
+:class:`PluginCallable` and :class:`PluginJob` respectively.
+
+.. seealso::
+
+    The :mod:`sopel.plugin`'s decorator create these objects directly.
+
+.. important::
+
+    This is all relatively new. Its usage and documentation is for Sopel core
+    development and advanced developers. It is subject to rapid changes
+    between versions without much (or any) warning.
+
+    Do **not** build your plugin based on what is here, you do **not** need to.
+
+"""
+# Copyright 2025, Florian Strzelecki <florian.strzelecki@gmail.com>
+#
+# Licensed under the Eiffel Forum License 2.
+from __future__ import annotations
+
+import abc
+import inspect
+import re
+from typing import (
+    Any,
+    Callable,
+    Literal,
+    Type,
+    TYPE_CHECKING,
+    TypeVar,
+)
+
+from typing_extensions import override
+
+from sopel.config.core_section import COMMAND_DEFAULT_HELP_PREFIX
+from sopel.lifecycle import deprecated
+
+
+if TYPE_CHECKING:
+    from sopel.bot import Sopel, SopelWrapper
+    from sopel.config import Config
+    from sopel.trigger import Trigger
+
+
+TypedPluginObject = TypeVar('TypedPluginObject', bound='AbstractPluginObject')
+"""A :class:`~typing.TypeVar` bound to :class:`AbstractPluginObject`.
+
+When used in the :meth:`AbstractPluginObject.from_plugin_object` class method,
+it means the return value must be an instance of the class used to call that
+method and not a different subclass of ``AbstractRule``.
+
+.. versionadded:: 8.1
+"""
+
+
+class AbstractPluginObject(abc.ABC):
+    """Abstract definition of a plugin object.
+
+    A plugin object encapsulates the logic and attributes required for a plugin
+    to register and execute this object.
+
+    .. versionadded:: 8.1
+    """
+    _sopel_callable = True
+
+    plugin_name: str | None
+    label: str | None
+    threaded: bool
+    doc: str | None
+
+    @property
+    @deprecated(
+        reason='`rule_label` is replaced by the `label` attribute.',
+        version='8.1',
+        removed_in='9.0',
+    )
+    def rule_label(self):
+        return self.label
+
+    @property
+    @deprecated(
+        reason='`thread` is replaced by the `threaded` attribute.',
+        version='8.1',
+        removed_in='9.0',
+    )
+    def thread(self):
+        # compatiblity property
+        # TODO: remove in Sopel 9
```

Comment:
I could fix the typo, or we could just rely on the `@deprecated` decorator to imply all of this 😁

```suggestion
```